### PR TITLE
[SPARK-10884] [ML] Support prediction on single instance for regression and classification related models

### PR DESCRIPTION
**Pull request by yanboliang**
Support prediction on single instance for regression and classification related models (i.e., PredictionModel, ClassificationModel and their sub classes).
Add corresponding test cases.

### CODE REVIEW

**Comment by yanboliang** [comment]
Jenkins, test this please.

File: mllib/src/main/scala/org/apache/spark/ml/classification/Classifier.scala [hunk 1]
```diff
@@ -129,12 +129,12 @@ abstract class ClassificationModel[FeaturesType, M <: ClassificationModel[Featur
 
   /**
    * Predict label for the given features.
-   * This internal method is used to implement [[transform()]] and output [[predictionCol]].
+   * This method can used to implement [[transform()]] and output [[predictionCol]].
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/classification/Classifier.scala]: nit: "This method can __be__ used..."

**Comment by yanboliang** [comment]
@mengxr @jkbradley Could you review this patch?

**Comment by jkbradley** [comment]
> Would transform and predict confuse users?

Probably.  Not sure what's best.  We could also just overload transform, but that would cause problems for the next issue you asked about.

> And what does predict mean by default if there could be multiple outputs, e.g., predicted class, confidence, etc? Do we have a default semantic that most users would agree on?

I think it's reasonable for "predict" to refer to the label being predicted.  Other outputs can have other more specific names.  (If you talk about the "prediction" for a prediction problem, it's pretty clear what you mean.)

**Comment by sethah** [comment]
ping @yanboliang Is this still on anyone's radar? I think it would be really useful. I can help review if needed. Maybe start by fixing merge conflicts? :D

**Comment by holdenk** [comment]
As we expose a new way to predict, maybe it would be good to also add a note that transform can be more efficient by broadcasting the model in certain cases or performing other optimizations, so as to encourage users to use the transform API when its the best path forward?

**Comment by fabiangebert** [comment]
do you know yet when this will be released?

**Comment by WeichenXu123** [comment]
I take this over in #19381

**Pull request closed without merging**